vm/rust/src/juno_state_reader.rs

```diff
@@ -311,68 +288,16 @@ fn native_try_from_json_string(
     raw_contract_class: &str,
     library_output_path: &PathBuf,
 ) -> Result<NativeContractClassV1, Box<dyn std::error::Error>> {
-    fn compile_and_load(
-        sierra_program: Program,
-        library_output_path: &PathBuf,
-    ) -> Result<AotNativeExecutor, Box<dyn std::error::Error>> {
-        let native_context = NativeContext::new();
-        let native_module = native_context.compile(&sierra_program)?;
-
-        persist_from_native_module(native_module, &sierra_program, library_output_path)
-    }
-
     let sierra_contract_class: cairo_lang_starknet_classes::contract_class::ContractClass =
         serde_json::from_str(raw_contract_class)?;
-
-    // todo(rodro): we are having two instances of a sierra program, one it's object form
-    // and another in its felt encoded form. This can be avoided by either:
-    //   1. Having access to the encoding/decoding functions
-    //   2. Refactoring the code on the Cairo mono-repo
-
     let sierra_program = sierra_contract_class.extract_sierra_program()?;
-
-    // todo(xrvdg) lift this match out of the function once we do not need sierra_program anymore
-    let executor = match load_compiled_contract(&sierra_program, library_output_path) {
-        Some(executor) => {
-            executor.or_else(|_err| compile_and_load(sierra_program, library_output_path))
-        }
-        None => compile_and_load(sierra_program, library_output_path),
-    }?;
-
-    Ok(NativeContractClassV1::new(executor, sierra_contract_class)?)
-}
-
-/// Load a contract that is already compiled.
-///
-/// Returns None if the contract does not exist at the output_path.
-///
-/// To compile and load a contract use [persist_from_native_module] instead.
-fn load_compiled_contract(
-    sierra_program: &Program,
-    library_output_path: &PathBuf,
-) -> Option<Result<AotNativeExecutor, Box<dyn std::error::Error>>> {
-    fn load(
-        sierra_program: &Program,
-        library_output_path: &PathBuf,
-    ) -> Result<AotNativeExecutor, Box<dyn std::error::Error>> {
-        let has_gas_builtin = sierra_program
-            .type_declarations
-            .iter()
-            .any(|decl| decl.long_id.generic_id.0.as_str() == "GasBuiltin");
-        let config = has_gas_builtin.then_some(Default::default());
-        let gas_metadata = GasMetadata::new(sierra_program, config)?;
-        let program_registry = ProgramRegistry::new(sierra_program)?;
-        let library = unsafe { Library::new(library_output_path)? };
-        Ok(AotNativeExecutor::new(
-            library,
-            program_registry,
-            gas_metadata,
-        ))
-    }
-
-    library_output_path
-        .is_file()
-        .then_some(load(sierra_program, library_output_path))
+    let executor = ContractExecutor::load(&library_output_path).unwrap_or({
+        let executor = ContractExecutor::new(&sierra_program, OptLevel::Default)?;
+        executor.save(&library_output_path)?;
+        executor
+    });
```

Comment:
This will always compile and save the contract as the argument to `unwrap_or` is eagerly evaluated. Use `unwrap_or_else` instead.

Reply:
Furthermore, `load` and `save` [needlessly borrow](https://rust-lang.github.io/rust-clippy/master/index.html#/needless_borrow) the `library_output_path`.

Reply:
Changed to `or_else` so I could still use "?" operator. Had to specify the type though since it couldn't figure it out.